### DISCUSSION TITLE
React Native atomWithStorage and createJSONStorage

### DISCUSSION THREAD

**Discussion by Xierruisi** [pmndrs/jotai#1429] [Q&A]
When I use these two functions, my page goes blank.No errors are reported.
 It's the same thing Async useAtom.
When I comment out Use Atom,The page is restored.

Atom.ts
``` ts
import { atomWithStorage, createJSONStorage } from 'jotai/utils';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { UserInfo } from '../service/User/types';

const tokenStorage = createJSONStorage<string>(() => AsyncStorage);
export const tokenAtom = atomWithStorage('token', '', tokenStorage);

const userInfoStorage = createJSONStorage<Partial<UserInfo>>(
    () => AsyncStorage,
);
export const userInfoAtom = atomWithStorage('userInfo', {}, userInfoStorage);

```

App.tsx
``` jsx
const App: React.FC = () => {
    return (
        <RootSiblingParent>
            <SafeAreaProvider>
                <Provider>
                    <Suspense fallback={<Text>Loading</Text>}>
                        <Routes />
                    </Suspense>
                </Provider>
            </SafeAreaProvider>
        </RootSiblingParent>
    );
};
```

Login.tsx
``` jsx
const Login: React.FC = () => {
    const navigation = useNavigation<StackNavigationProp<RootNavigationList>>();
    const [phoneOrEmail, setPhoneOrEmail] = useState<string>('');
    const [password, setPassword] = useState<string>('');
    const [, setToken] = useAtom(tokenAtom);
    const [, setUserInfo] = useAtom(userInfoAtom);
    ....
    return (
        <View style={styles.LoginScreenContainer}>
            <Icon
                name="alpha-f-box"
                type="material-community"
                color={'#2089dc'}
                size={104}
            />
       ...

}
```

**Comment by dai-shi** [comment]
Can it be reproduced with just one atom? For example, `userInfoAtom` only?

**Reply by Xierruisi** [reply]
No,when i use asyncAtom in my components ,the components goes blank

**Reply by Xierruisi** [reply]
I tried a different approach to Implement my business. But I got an error.
``` ts
    useEffect(() => {
        const getToken = async () => {
            const token = await AsyncStorage.getItem('token');
            return token || '';
        };
        const getUserInfo = async () => {
            let parseUserInfo = {};
            const userInfo = await AsyncStorage.getItem('uesrInfo');
            userInfo && (parseUserInfo = JSON.parse(userInfo) as UserInfo);
            return parseUserInfo;
        };
        // @ts-ignore
        setUserInfo(getUserInfo());
        // @ts-ignore
        setToken(getToken());
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);
```

``` 
Error: A component suspended while responding to synchronous input. This will cause the UI to be replaced with a loading indicator. To fix, updates that suspend should be wrapped with startTransition.

This error is located at:
    in App
    in RCTView (created by View)
    in View (created by AppContainer)
    in RCTView (created by View)
    in View (created by AppContainer)
    in AppContainer
    in Fixon(RootComponent), js engine: hermes
```

<img width="404" alt="image" src="https://user-images.githubusercontent.com/51395487/191065485-acc61afa-295a-48a5-b970-8eba3a4768b2.png">

**Reply by Xierruisi** [reply]
pakage.json
``` json
{
        "@react-native-async-storage/async-storage": "^1.17.10",
        "@react-native-community/datetimepicker": "^6.3.3",
        "@react-navigation/bottom-tabs": "^6.3.3",
        "@react-navigation/drawer": "^6.4.4",
        "@react-navigation/native": "^6.0.12",
        "@react-navigation/stack": "^6.2.3",
        "@rneui/base": "^4.0.0-rc.6",
        "@rneui/themed": "^4.0.0-rc.6",
        "@types/lodash": "^4.14.185",
        "axios": "^0.27.2",
        "jotai": "^1.8.3",
        "lodash": "^4.17.21",
        "react": "18.1.0",
        "react-native": "0.70.0",
        "react-native-calendars": "^1.1289.0",
        "react-native-gesture-handler": "^2.6.0",
        "react-native-reanimated": "^2.10.0",
        "react-native-root-siblings": "^4.1.1",
        "react-native-root-toast": "^3.4.0",
        "react-native-safe-area-context": "^4.3.3",
        "react-native-screens": "^3.17.0",
        "react-native-ui-lib": "^6.21.2",
        "react-native-vector-icons": "^9.2.0"
}
```

**Comment by dai-shi** [comment]
Does having just one custom storage help anything?
```js
import { atomWithStorage, createJSONStorage } from 'jotai/utils';
import AsyncStorage from '@react-native-async-storage/async-storage';

const customStorage = createJSONStorage(() => AsyncStorage);

export const tokenAtom = atomWithStorage('token', '', customStorage);

export const userInfoAtom = atomWithStorage('userInfo', {}, customStorage);
```

**Reply by Xierruisi** [reply]
> Does having just one custom storage help anything?
> 
> ```js
> import { atomWithStorage, createJSONStorage } from 'jotai/utils';
> import AsyncStorage from '@react-native-async-storage/async-storage';
> 
> const customStorage = createJSONStorage(() => AsyncStorage);
> 
> export const tokenAtom = atomWithStorage('token', '', customStorage);
> 
> export const userInfoAtom = atomWithStorage('userInfo', {}, customStorage);
> ```


Nope, It still keep empty screen.

**Reply by smontlouis** [reply]
@MircoSans did you fix the issue ? I'm using asyncStorage + suspense, and when importing useAtom(asyncAtom), screen get stuck in suspense fallback

**Reply by smontlouis** [reply]
@dai-shi any news regarding this issue ?

**Reply by dai-shi** [reply]
Can you guys help identifing what the issue is? Is it specific to RN AsyncStorage? Or, is it the issue with atomWithStorage and can be reproduced with web localStorage?

fwiw, in v2 #1515, we will make `delayInit: true` by default to avoid various issues. So, please specify `delayInit: true` even for now.

**Reply by smontlouis** [reply]
I solved it by make delayInit to true thank you !

**Reply by ansh** [reply]
This only started happening in Expo v47 for me. In v45 and v46 it was totally fine and there was no issue with the screen going blank. 

Expo v47 introduced React 18.1 and React Native 0.70.5

I believe the issue might be because of that. @dai-shi Thoughts?

edit:
I checked the docs https://jotai.org/docs/guides/persistence and implemented the `<Preload />` component. It seems to resolve the issue without using `delayInit: true` 
However, the app still crashes when doing a full-reload using the shortcut `R` (Look at keyboard shortcut for `R` key https://docs.expo.dev/workflow/expo-cli/). However, if reloaded from the dev-menu (`cmd+ctrl+Z` or `cmd+D` and then pressing the "Reload" button on screen), it is fine.

Not sure why this is happening but seems like the culprit is Suspense.

**Reply by dai-shi** [reply]
Not sure. Did you try `delayInit: true`?

**Reply by ansh** [reply]
Yes, I tried `delayInit: true` and it also seems to resolve the issue (of the app displaying a blank screen) just like the `<Preload />` component does.

However, both solutions still have the reload issue that I described above.

**Reply by ansh** [reply]
@dai-shi It's quite an odd behavior. The "reload issue" goes away if I change the value of the stored item, and then reload. 

This is reproducible 100% of the time. I can share a link to a new project with this issue if it will help you fix it/you have time to fix it.

**Reply by dai-shi** [reply]
If there's a bug with `delayInit: true`, we should investigate to fix it. I would need a reproduction without RN, ideally with one based on https://csb.jotai.org.

Do you think it's specific to Expo 47?

**Reply by ansh** [reply]
@dai-shi I am pretty confident it is specific to Expo 47. This issue only started occurring after I upgraded from Expo 46 to 47 today. Before this, it was not a problem.

Not sure how I can reproduce it without React Native as it seems it is a React Native specific issue. I’ll create a simple Expo project to reproduce it and hopefully that will help you.

**Reply by ansh** [reply]
@dai-shi Nevermind, this crash is caused by Reanimated library and is specific to Expo 47 as it is due to Fabric. Jotai isn't doing anything.

**Reply by nksfrank** [reply]
I'm facing this issue as well and I'm not using Expo for my RN app
